src/NLSModels.jl

```diff
@@ -93,8 +93,8 @@ include("feasibility_residual.jl")
 
 Computes F(x), the residual at x.
 """
-function residual(nls :: AbstractNLSModel, x :: AbstractVector)
-  Fx = zeros(nls_meta(nls).nequ)
+function residual(nls :: AbstractNLSModel, x :: AbstractVector{T}) where T <: Real
+  Fx = zeros(T, nls_meta(nls).nequ)
   residual!(nls, x, Fx)
```

Comment:
Here it seems to me that we make the assumption that problem functions (perhaps defined by the user) are able to work with various types. That's reasonable for Julia functions and ForwardDiff models (and hopefully JuMP soon). But what about others, such as CUTEstModels? I suppose it's ok to just let them raise an error?!

Reply:
This heavily relies on `x`:
- If `residual!` works for `x` with `eltype=T`, then `residual` is well defined.
- If `residual!` doesn't for `x` with `eltype=T`, then `residual` doesn't work as well.